integration/integration_test.go

```diff
@@ -0,0 +1,230 @@
+
+	// output should only have events with pids of 1
+	pids := strings.Split(strings.TrimSpace(gotOutput.String()), "\n")
+	for _, p := range pids {
+		pid, _ := strconv.Atoi(p)
+		require.Equal(t, 1, pid)
+	}
+}
+
+// check that execve event is called
+func checkExecve(t *testing.T, gotOutput *bytes.Buffer) {
+	_, _ = exec.Command("ls").CombinedOutput()
+
+	// check output length
+	require.NotEmpty(t, gotOutput.String())
+
+	// output should only have events with event name of execve
+	eventNames := strings.Split(strings.TrimSpace(gotOutput.String()), "\n")
+	for _, en := range eventNames {
+		if len(en) > 0 {
+			require.Equal(t, "execve", en)
+		}
+	}
+}
+
+// check for filesystem set when ls is invoked
+func checkSetFs(t *testing.T, gotOutput *bytes.Buffer) {
+	_, _ = exec.Command("ls").CombinedOutput()
+
+	// check output length
+	require.NotEmpty(t, gotOutput.String())
+
+	// output should only have events with event name of execve
+	eventNames := strings.Split(strings.TrimSpace(gotOutput.String()), "\n")
+	for _, en := range eventNames {
+		require.Contains(t, []string{"access", "openat", "fstat", "close", "read", "pread64", "statfs", "ioctl", "statx", "getdents64", "write"}, en)
+	}
+}
+
+func Test_Events(t *testing.T) {
+	var testCases = []struct {
+		name      string
+		args      []string
+		eventFunc func(*testing.T, *bytes.Buffer)
+	}{
+		{
+			name:      "do a file write",
+			args:      []string{"--trace", "event=magic_write"},
+			eventFunc: checkMagicwrite,
+		},
+		{
+			name:      "execute a command",
+			args:      []string{"--trace", "comm=ls", "--output", "gotemplate=processName.tmpl"},
+			eventFunc: checkExeccommand,
+		},
+		{
+			name:      "trace new pids",
+			args:      []string{"--trace", "pid=new", "--output", "gotemplate=pid.tmpl"},
+			eventFunc: checkPidnew,
+		},
+		{
+			name:      "trace uid 0 with comm ls",
+			args:      []string{"--trace", "uid=0", "--trace", "comm=ls", "--output", "gotemplate=uid.tmpl"},
+			eventFunc: checkUidzero,
+		},
+		{
+			name:      "trace pid 1",
+			args:      []string{"--trace", "pid=1", "--output", "gotemplate=pid.tmpl"},
+			eventFunc: checkPidOne,
+		},
+		// TODO: Add pid=0,1
+		// TODO: Add pid=0 pid=1
+		// TODO: Add uid>0
+		// TODO: Add pid>0 pid<1000
+		// TODO: Add u>0 u!=1000
+		{
+			name:      "trace only execve events from comm ls",
+			args:      []string{"--trace", "event=execve", "--output", "gotemplate=eventName.tmpl"},
+			eventFunc: checkExecve,
+		},
+		{
+			name:      "trace filesystem events from comm ls",
+			args:      []string{"--trace", "s=fs", "--trace", "comm=ls", "--output", "gotemplate=eventName.tmpl"},
+			eventFunc: checkSetFs,
+		},
+	}
+
+	checkTraceeBinary(t)
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(st *testing.T) {
+			st.Parallel()
+
+			var gotOutput bytes.Buffer
+			done := make(chan bool, 1)
+			go loadTracee(st, &gotOutput, done, tc.args...)
+			time.Sleep(time.Second) // wait for tracee init
```

Comment:
the more accurate way is to look for the file `$outputdir/tracee.pid`

Reply:
That's interesting. I had a look at it but found an improvement we can make before we can use it. Today the code looks like was written to only run a single instance of Tracee at a time. https://github.com/aquasecurity/tracee/blob/228c6d329e86dd8f62bb5fe48761e78b67dafb39/tracee-ebpf/tracee/tracee.go#L328-L331

But this has changed now so we will need to support creating a unique pid file for each run. `WriteFile` truncates the file if it exists and rewrites it. This prevents us from relying on it as a readiness check across multiple goroutines. But I do agree with you this is the better approach.

Reply:
Filed an issue here to track it https://github.com/aquasecurity/tracee/issues/607